apps/rush-lib/src/logic/taskRunner/BaseBuilder.ts

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { ITaskWriter } from '@rushstack/stream-collator';
+import { TaskStatus } from './TaskStatus';
+
+/**
+ * The `Task` class is a node in the dependency graph of work that needs to be scheduled by the `TaskRunner`.
+ * Each `Task` has a `BaseBuilder` member, whose subclass manages the actual operations for building a single
+ * project.
+ */
+export abstract class BaseBuilder {
+  /**
+   * Name of the task definition.
+   */
+  abstract name: string;
+
+  /**
+   * This flag determines if an incremental build is allowed for the task.
+   */
+  abstract isIncrementalBuildAllowed: boolean;
+
+  /**
+   * Assigned by execute().  True if the build script was an empty string.  Operationally an empty string is
+   * like a shell command that succeeds instantly, but e.g. it would be odd to report build time statistics for it.
+   */
+  abstract hadEmptyScript: boolean;
```

Comment:
```suggestion
  /**
   * Name of the task definition.
   */
  abstract readonly name: string;

  /**
   * This flag determines if an incremental build is allowed for the task.
   */
  abstract readonly isIncrementalBuildAllowed: boolean;

  /**
   * Assigned by execute().  True if the build script was an empty string.  Operationally an empty string is
   * like a shell command that succeeds instantly, but e.g. it would be odd to report build time statistics for it.
   */
  abstract readonly hadEmptyScript: boolean;
```

Reply:
`name` can be `readonly`.  The other properties are mutable.

The `BaseBuilder` state management needs to be redesigned -- it's just a transformation of the `ITaskDefinition` interface which had no private/public separation.  But I don't have time to get into that in this PR.